pkg/k8s/resource/diffstore.go

```diff
@@ -0,0 +1,270 @@
+	dsf.onDiff = append(dsf.onDiff, newOnDiff)
+	done = func() {
+		// Close
+		close(newOnDiff)
+		// And drain
+		select {
+		case <-newOnDiff:
+		default:
+		}
+	}
+	return newOnDiff, done
+}
+
+// Start implements hive.Hook
+func (dsf *diffStoreFactory[T]) Start(_ hive.HookContext) error {
+	go dsf.run()
+	return nil
+}
+
+// Stop implements hive.Hook
+func (dsf *diffStoreFactory[T]) Stop(stopCtx hive.HookContext) error {
+	dsf.cancel()
+
+	select {
+	case <-dsf.doneChan:
+	case <-stopCtx.Done():
+	}
+
+	return nil
+}
+
+func (dsf *diffStoreFactory[T]) run() {
+	defer close(dsf.doneChan)
+
+	var err error
+	dsf.store, err = dsf.params.Resource.Store(dsf.ctx)
+	if err != nil {
+		dsf.params.Shutdowner.Shutdown(hive.ShutdownWithError(err))
+		return
+	}
+
+	dsf.mu.Lock()
+	for _, store := range dsf.diffStores {
+		store.store = dsf.store
+	}
+	dsf.mu.Unlock()
+
+	errChan := make(chan error, 2)
+	updateChan := stream.ToChannel[Event[T]](dsf.ctx, errChan, dsf.params.Resource)
+
+	for {
+		select {
+		case event, ok := <-updateChan:
+			if !ok {
+				break
+			}
+
+			event.Handle(
+				func() error {
+					dsf.inSync = true
+					dsf.signalObservers()
+					return nil
+				},
+				dsf.handleChange,
+				dsf.handleChange,
+			)
+		case err := <-errChan:
+			dsf.params.Shutdowner.Shutdown(hive.ShutdownWithError(err))
+			return
+		}
+	}
+}
+
+func (dsf *diffStoreFactory[T]) handleChange(k Key, _ T) error {
+	dsf.mu.Lock()
+	defer dsf.mu.Unlock()
+
+	for _, store := range dsf.diffStores {
+		store.markChanged(k)
+	}
+
+	if dsf.inSync {
+		dsf.signalObservers()
+	}
+
+	return nil
+}
+
+func (dsf *diffStoreFactory[T]) signalObservers() {
+	for _, onDiff := range dsf.onDiff {
+		select {
+		case onDiff <- struct{}{}:
+		default:
+		}
+	}
+}
+
+// DiffStore is a super set of the Store. The diffStore tracks all changes made to the diff store since the
+// last time the user synced up. This allows a user to get a list of just the changed objects while still being able
+// to query the full store for a full sync.
+type DiffStore[T k8sRuntime.Object] interface {
+	Store[T]
+
+	// Diff returns a list of items that have been upserted(updated or inserted) and deleted since the last call to Diff.
+	Diff() (upserted []T, deleted []Key, err error)
+	// MarkChanged marks a key as having been changed so it will appear in the Diff again. This can be used in case
+	// of a failure to process so a next call to Diff gets the chance to process the key again.
+	MarkChanged(Key)
+}
+
+var _ DiffStore[*k8sRuntime.Unknown] = (*diffStore[*k8sRuntime.Unknown])(nil)
+
+// diffStore takes a Resource[T] and watches for events, it stores all of the keys that have been changed.
+// diffStore can still be used as a normal store, but adds the Diff function to get a Diff of all changes.
+// The diffStore also takes in Signaler which it will signal after the initial sync and every update thereafter.
+type diffStore[T k8sRuntime.Object] struct {
+	store Store[T]
```

Comment:
You could just embed `Store[T]` directly as then you don't need to implement wrappers for `List()` etc.